integrations/git_annex_test.go

```diff
@@ -0,0 +1,940 @@
+	if err != nil {
+		return err
+	}
+
+	// verify the file was uploaded
+	localObjectPath, err := annexObjectPath(repoPath, "contribution.bin")
+	if err != nil {
+		return err
+	}
+	//localObjectPath := path.Join(repoPath, "contribution.bin") // or, just compare against the checked-out file
+
+	remoteObjectPath, err := annexObjectPath(remoteRepoPath, "contribution.bin")
+	if err != nil {
+		return err
+	}
+
+	match, err := filecmp(localObjectPath, remoteObjectPath, 0)
+	if err != nil {
+		return err
+	}
+	if !match {
+		return errors.New("Annexed files should be the same")
+	}
+
+	return nil
+}
+
+// ---- Helpers ----
+
+func generateRandomFile(size int, path string) (err error) {
+	// Generate random file
+
+	// XXX TODO: maybe this should not be random, but instead a predictable pattern, so that the test is deterministic
+	bufSize := 4 * 1024
+	if bufSize > size {
+		bufSize = size
+	}
+
+	buffer := make([]byte, bufSize)
+
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	written := 0
+	for written < size {
+		n := size - written
+		if n > bufSize {
+			n = bufSize
+		}
+		_, err := rand.Read(buffer[:n])
+		if err != nil {
+			return err
+		}
+		n, err = f.Write(buffer[:n])
+		if err != nil {
+			return err
+		}
+		written += n
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// https://stackoverflow.com/a/30038571
+func filecmp(file1, file2 string, chunkSize int) (bool, error) {
+	// Check file size ...
+	if chunkSize == 0 {
+		chunkSize = 2 << 12
+	}
+
+	f1, err := os.Open(file1)
+	if err != nil {
+		return false, err
+	}
+	defer f1.Close()
+
+	f2, err := os.Open(file2)
+	if err != nil {
+		return false, err
+	}
+	defer f2.Close()
+
+	for {
+		b1 := make([]byte, chunkSize)
+		_, err1 := f1.Read(b1)
+		if err1 != nil && err1 != io.EOF {
+			return false, err1
+		}
+
+		b2 := make([]byte, chunkSize)
+		_, err2 := f2.Read(b2)
+		if err2 != nil && err2 != io.EOF {
+			return false, err2
+		}
+
+		if err1 == io.EOF && err2 == io.EOF {
+			return true, nil
+		} else if err1 != nil || err2 != nil {
+			return false, nil
+		}
+
+		if !bytes.Equal(b1, b2) {
+			return false, nil
+		}
+	}
+}
+
+// ---- Annex-specific helpers ----
+
+/* Initialize a repo with some baseline annexed and non-annexed files.
+
+TODO: this could be replaced with a fixture repo;
+see integrations/gitea-repositories-meta/ and models/fixtures/repository.yml.
+
+However we reuse this template for -different- repos.
+*/
+func doInitAnnexRepository(t *testing.T, repoPath string) {
+	// set up what files should be annexed
+	// in this case, all *.bin  files will be annexed
+	// without this, git-annex's default config annexes every file larger than some number of megabytes
+	f, err := os.Create(path.Join(repoPath, ".gitattributes"))
+	require.NoError(t, err)
+	f.WriteString("*                   annex.largefiles=nothing")
+	f.WriteString("*.bin  filter=annex annex.largefiles=anything")
```

Comment:
I'd prefer to `require.NoError` for these writes as well? No need to worry about partial writes, because those will have a non-nil error.

Reply:
Yes you're right!

Actually I think I'm going to rewrite this to return an `error` instead of passing in `t`. I started doing that halfway and only using `require.whatever(t, ...)` inside the `t.Run()`s, but this was one of the earlier subroutines I wrote. That was the way the Gitea authors wrote `git_tests.go`, but it's a lot more flexible if the helpers don't think about the tests.